src/bin/ion/commands/beta/generate/model.rs

```diff
@@ -0,0 +1,154 @@
+// Also, current approach doesn't allow having nested sequences in the generated code. Because the `element_type` in `AbstractDataType::Sequence`
+// doesn't have information on its nested types' `element_type`. This can be resolved with below defined new data model.
+// _Note: This model will eventually use a map (FullQualifiedTypeReference, DataModel) to resolve some the references in container types(sequence or structure)._
+// TODO: This is not yet used in the implementation, modify current implementation to use this data model.
+use crate::commands::beta::generate::context::SequenceType;
+use serde::Serialize;
+
+#[allow(dead_code)]
+#[derive(Debug, Clone, PartialEq)]
+pub struct DataModel {
+    // Represents the fully qualified name for this data model
+    name: FullyQualifiedTypeName,
+    // Represents doc comment for the generated code
+    doc_comment: String,
+    // Represents the type of the data model
+    // It can be `None` for modules or packages.
+    code_gen_type: Option<Box<CodeGenType>>,
+    // Represents the nested types for this data model
+    nested_types: Vec<DataModel>,
+    // Represents the source ISL type which can be used to get other constraints useful for this type.
+    // For example, getting the length of this sequence from `container_length` constraint or getting a `regex` value for string type.
+    // This will also be useful for `text` type to verify if this is a `string` or `symbol`.
+    source: Option<IslType>,
+}
+
+impl DataModel {
+    pub fn is_scalar(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type.deref(), CodeGenType::Scalar);
+        }
+        false
+    }
+
+    pub fn is_sequence(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type.deref(), CodeGenType::Sequence(_));
+        }
+        false
+    }
+
+    pub fn is_structure(&self) -> bool {
+        if let Some(code_gen_type) = &self.code_gen_type {
+            return matches!(code_gen_type.deref(), CodeGenType::Structure(_));
+        }
+        false
+    }
+}
+
+/// Represents a fully qualified type name for a type definition
+/// e.g. For a `Foo` class in `org.example` namespace
+///     In Java, `org.example.Foo`
+///     In Rust, `org::example::Foo`
+type FullyQualifiedTypeName = Vec<String>;
+
+/// Represents a fully qualified type name for a type reference
+#[allow(dead_code)]
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct FullQualifiedTypeReference {
+    // Represents fully qualified name of the type
+    // e.g. In Java, `org.example.Foo`
+    //      In Rust, `crate::org::example::Foo`
+    type_name: FullyQualifiedTypeName,
+    // For types with parameters this will represent the nested parameter
+    parameters: Box<FullQualifiedTypeReference>,
```

Comment:
Should this be `Option<Box<_>>`?

Reply:
My bad! I previously was using the model itself instead of using `FullyQualifiedName` which is why I used `Box`. can remove it now.